pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,226 @@
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/btree"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	ktypes "k8s.io/apimachinery/pkg/types"
+
+	"github.com/openshift/origin/pkg/oauth/apis/oauth"
+	oauthclient "github.com/openshift/origin/pkg/oauth/generated/internalclientset/typed/oauth/internalversion"
+	oauthlister "github.com/openshift/origin/pkg/oauth/generated/listers/oauth/internalversion"
+)
+
+var ErrTimedout = errors.New("Token timed out")
+
+type tokenData struct {
+	name     string
+	client   string
+	seen     time.Time
+	creation time.Time
+	timeout  int32
+}
+
+type tokenDataRef struct {
+	name    string
+	timeout time.Time
+}
+
+func (a *tokenDataRef) Less(than btree.Item) bool {
+
+	if a.timeout.Equal(than.(*tokenDataRef).timeout) {
+		return a.name < than.(*tokenDataRef).name
+	}
+	return a.timeout.Before(than.(*tokenDataRef).timeout)
+}
+
+type TokenTimeoutUpdater struct {
+	oauthClient    oauthlister.OAuthClientLister
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan tokenData
+	data           map[string]tokenData
+	tree           *btree.BTree
+	defaultTimeout time.Duration
+	flushTimeout   time.Duration
+	safetyMargin   time.Duration
+}
+
+const (
+	// we never flush more frequently than every 2.5 minutes
+	// except for emergency flushouts
+	minFlushTimeout int32 = 150
+)
+
+func NewTokenTimeoutUpdater(tokens oauthclient.OAuthAccessTokenInterface, oauthClient oauthlister.OAuthClientLister, defaultTimeout int32) *TokenTimeoutUpdater {
+	if defaultTimeout < minFlushTimeout {
+		glog.Errorf("Default Min Timeout is set at %d seconds. "+
```

Comment:
Validate in master config.

Reply:
Meaning ?
I do not want to fail to start because tests will set this smaller. Also I want this just to be a warning, you cans et a shorter timeout it will just cause every auth to cause an emergency flush

Reply:
I changed this to an Infof and a "not recommended" message

Reply:
If you assign an invalid config, the master must fail to start.  We do not "warn" about invalid config.  We refuse to start under such situations.

Reply:
It is not invalid, it is just not recommended, as it will cause a lot of churn.